{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "JQnxbTZERfbi",
      "metadata": {
        "id": "JQnxbTZERfbi",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "#Python Runtime\n",
        "from google.colab import drive\n",
        "drive.mount('/content/drive')\n",
        "\n",
        "#system(\"tar xf /content/drive/MyDrive/Colab Notebooks/lib/library.tar.gz\")\n",
        "#.libPaths('/content/usr/local/lib/R/site-library/')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "E5KBkpmJDkYQ",
      "metadata": {
        "id": "E5KBkpmJDkYQ",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "#R Runtime\n",
        "install.packages(\"ggplot2\")\n",
        "install.packages(\"dplyr\")\n",
        "#system(\"apt-get install gfortran\")\n",
        "#install.packages(\"ggpubr\")\n",
        "install.packages(\"tidyr\")\n",
        "devtools::install_github(\"caleblareau/BuenColors\")\n",
        "install.packages('patchwork')\n",
        "install.packages('cowplot')\n",
        "devtools::install_github(\"jakelawlor/PNWColors\")\n",
        "\n",
        "library(PNWColors)\n",
        "library(tidyr)\n",
        "library(ggplot2)\n",
        "library(dplyr)\n",
        "#library(ggpubr)\n",
        "library(BuenColors)\n",
        "library(cowplot)\n",
        "library(patchwork)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "NmAf1adsbXtp",
      "metadata": {
        "id": "NmAf1adsbXtp",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "#tar(\"library.tar.gz\", \"/usr/local/lib/R/site-library\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "1HzNSEMGEJO2",
      "metadata": {
        "id": "1HzNSEMGEJO2",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "data_path=\"/content/drive/MyDrive/Manuscripts/2023_UKBB-BBJ-GTEx_MPRA/2023_UKBB_GTEx_MPRA_Manuscript/\"\n",
        "output_path=\"/content/drive/MyDrive/Temp/\"\n",
        "\n",
        "gtex_mpra_paired_final20230117 <- read.delim(paste0(data_path,\"Data/MPRA/gtex_mpra_final20230117.txt.gz\"),header=TRUE,stringsAsFactors = FALSE,check.names=FALSE)\n",
        "traits_mpra_paired_final20230117 <- read.delim(paste0(data_path,\"Data/MPRA/traits_mpra_final20230117.txt.gz\"),header=TRUE,stringsAsFactors = FALSE,check.names=FALSE)\n",
        "haplotypes_data <- read.table(paste0(data_path,\"Data/MPRA/haplos_merged_20240202.txt\"), sep = '\\t', header = TRUE)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "EOJ75MG2bLHD",
      "metadata": {
        "id": "EOJ75MG2bLHD",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "if (!dir.exists(paste0(output_path,\"plots/\"))) {\n",
        "  dir.create(paste0(output_path,\"plots/\"))\n",
        "} else {\n",
        "  cat(\"Directory already exists!\\n\")\n",
        "}\n",
        "\n",
        "if (!dir.exists(paste0(output_path,\"haplo_plots/\"))) {\n",
        "  dir.create(paste0(output_path,\"haplo_plots/\"))\n",
        "} else {\n",
        "  cat(\"Directory already exists!\\n\")\n",
        "}"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "e-QUMvuXLeO2",
      "metadata": {
        "id": "e-QUMvuXLeO2",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "desired_ids <- c(\"chr12:31226835:T:A\", \"chr13:50432800:T:C\", \"chr14:65175364:A:C\", \"chr15:55526886:C:T\", \"chr16:193224:G:A\", \"chr19:58662235:C:T\", \"chr19:58662254:T:C\", \"chr2:191049403:G:A\", \"chr20:45944659:G:C\", \"chr20:5485337:G:A\", \"chr20:61464184:C:T\", \"chr21:37572401:T:G\", \"chr22:19132380:A:T\", \"chr22:19132385:G:A\", \"chr22:50742721:T:A\", \"chr3:42306446:C:T\", \"chr5:95163449:G:A\", \"chr6:7143859:C:A\", \"chr7:105416560:C:A\", \"chr7:139875230:C:T\")\n",
        "\n",
        "subset_gtex <- gtex_mpra_paired_final20230117 %>% filter(variant %in% desired_ids)\n",
        "\n",
        "glimpse(subset_gtex)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "dGrPyAlOYRpi",
      "metadata": {
        "id": "dGrPyAlOYRpi",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "selected_columns <- c(\"variant\",\"cell_type\",\"A_log2FC\", \"B_log2FC\", \"A_log2FC_SE\", \"B_log2FC_SE\")  # Replace with actual column names\n",
        "\n",
        "pre_filtered_data <- subset_gtex %>%\n",
        "  dplyr::select(all_of(selected_columns))\n",
        "\n",
        "#plot_data_melted <- pre_filtered_data %>%\n",
        "#  pivot_longer(\n",
        "#    cols = c(\"A_log2FC\", \"B_log2FC\"),\n",
        "#    names_to = \"log2FC_type\",\n",
        "#    values_to = \"log2FC\"\n",
        "#  )\n",
        "\n",
        "  melted_data <- pre_filtered_data %>%\n",
        "  pivot_longer(\n",
        "    cols = c(A_log2FC, A_log2FC_SE, B_log2FC, B_log2FC_SE),\n",
        "    names_to = c(\"set\", \".value\"),\n",
        "    names_pattern = \"(A|B)_(.*)\"\n",
        "  ) %>%\n",
        "  distinct()\n",
        "\n",
        "#melted_data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "fifSsSmd4JwA",
      "metadata": {
        "id": "fifSsSmd4JwA",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "# Setup theme\n",
        "my_theme <-\n",
        "  BuenColors::pretty_plot(fontsize = 8) +\n",
        "  BuenColors::L_border() +\n",
        "  theme(\n",
        "    plot.background = element_blank(),\n",
        "    plot.margin = margin(0, 0.1, 0, 0.1, unit = \"cm\"),\n",
        "    plot.title = element_text(hjust = 4e-3, margin = margin(b = -12)),\n",
        "    # legend.position = \"none\",\n",
        "    legend.position = c(0.9, 1),\n",
        "    legend.justification = c(1, 1),\n",
        "    legend.title = element_text(margin = margin(0, 0, 0, 0)),\n",
        "    legend.background = element_blank(),\n",
        "    legend.key.size = unit(0.2, \"cm\")\n",
        "\n",
        "  )"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "M72X0AMbl-54",
      "metadata": {
        "id": "M72X0AMbl-54",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "\n",
        "\n",
        "\n",
        "font_family <- \"Arial\"\n",
        "title_size <- 16\n",
        "axis_title_size <- 16\n",
        "axis_text_size <- 12\n",
        "legend_title_size <- 16\n",
        "legend_text_size <- 12\n",
        "legend_key_size <- 1\n",
        "\n",
        "clrs = c(\n",
        "  pnw_palette('Bay',8,type='continuous')[5],\n",
        "  pnw_palette('Bay',8,type='continuous')[3],\n",
        "  pnw_palette('Starfish')[5],\n",
        "  pnw_palette('Bay',8,type='continuous')[1],\n",
        "  pnw_palette('Bay',8,type='continuous')[8],\n",
        "  'black'\n",
        ")\n",
        "\n",
        "names(clrs) = c(\"A549\", \"HCT116\", \"HEPG2\", \"K562\",\"SKNSH\",\"any\")\n",
        "base_colors<-clrs\n",
        "\n",
        "lighten_color <- function(color, amount = 0.3) {\n",
        "  alpha(color, amount)\n",
        "}\n",
        "\n",
        "color_palette <- c()\n",
        "for (cell_type in names(base_colors)) {\n",
        "  color_palette[paste0(cell_type, \"_A\")] <- base_colors[cell_type]\n",
        "  color_palette[paste0(cell_type, \"_B\")] <- lighten_color(base_colors[cell_type])\n",
        "}\n",
        "\n",
        "uniq_ID=unique(melted_data$variant)\n",
        "\n",
        "for (i in 1:length(uniq_ID)) {\n",
        "\n",
        "cur_ID=uniq_ID[i]\n",
        "print(i)\n",
        "print(cur_ID)\n",
        "\n",
        "filtered_data <- melted_data %>%\n",
        "  filter(variant == cur_ID)\n",
        "\n",
        "#print(filtered_data)\n",
        "\n",
        "filtered_data$label <- ifelse(filtered_data$set == \"A\", \"Ref\", \"Alt\")\n",
        "\n",
        "filtered_data$cell_type_set <- interaction(filtered_data$cell_type, filtered_data$set, sep = \"_\")\n",
        "filtered_data$cell_type_set <- factor(filtered_data$cell_type_set, levels = levels(filtered_data$cell_type_set))\n",
        "\n",
        "\n",
        "p1<-ggplot(filtered_data, aes(x = cell_type, y = log2FC, fill = cell_type_set)) +\n",
        "  geom_bar(stat = \"identity\", position = position_dodge(width = 0.8), width = 0.7) +\n",
        "  geom_errorbar(aes(ymin = log2FC - log2FC_SE, ymax = log2FC + log2FC_SE),\n",
        "                position = position_dodge(width = 0.8), width = 0) +\n",
        "  geom_text(aes(y=-0, label = label), size=5, position = position_dodge(width = 0.8),\n",
        "            angle = 0, vjust = 0) +\n",
        "  labs(y = \"Activity - Log2(RNA/DNA)\", x = \"Cell Type\") +\n",
        "  pretty_plot() +\n",
        "  my_theme +\n",
        "  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +  # Adjust x-axis limits and expansion\n",
        "  scale_x_discrete(expand = expansion(add = c(0.5, 0.5))) + # Reduce white space above and below bars\n",
        "  theme(\n",
        "          text = element_text(family = font_family),  # Set font family for the entire plot\n",
        "          plot.title = element_text(size = title_size),\n",
        "          axis.title = element_text(size = axis_title_size),\n",
        "          axis.text = element_text(size = axis_text_size),\n",
        "          axis.text.x = element_text(angle = 0, hjust=0.5),\n",
        "          axis.title.x = element_text(angle=0, hjust=0.5)\n",
        "    ) +\n",
        "  scale_fill_manual(values = color_palette) +\n",
        "  theme(legend.position = \"none\")\n",
        "\n",
        "filtered_data <- melted_data %>%\n",
        "  filter(variant == cur_ID)\n",
        "\n",
        "filtered_data <- filtered_data %>%\n",
        "  filter(cell_type %in% c(\"HEPG2\", \"K562\"))\n",
        "\n",
        "filtered_data$label <- ifelse(filtered_data$set == \"A\", \"Ref\", \"Alt\")\n",
        "filtered_data$label <- factor(filtered_data$label, levels = c(\"Ref\", \"Alt\"))\n",
        "\n",
        "filtered_data$cell_type_set <- interaction(filtered_data$cell_type, filtered_data$set, sep = \"_\")\n",
        "filtered_data$cell_type_set <- factor(filtered_data$cell_type_set, levels = levels(filtered_data$cell_type_set))\n",
        "\n",
        "p2<-ggplot(filtered_data, aes(x = cell_type, y = log2FC, fill = cell_type_set)) +\n",
        "  geom_bar(stat = \"identity\", position = position_dodge(width = 0.8), width = 0.7) +\n",
        "  geom_errorbar(aes(ymin = log2FC - log2FC_SE, ymax = log2FC + log2FC_SE),\n",
        "                position = position_dodge(width = 0.8), width = 0) +\n",
        "  geom_text(aes(y=-0, label = label), size=5, position = position_dodge(width = 0.8),\n",
        "            angle = 0, vjust = 0) +\n",
        "  labs(y = \"Activity - Log2(RNA/DNA)\", x = \"Cell Type\") +\n",
        "  pretty_plot() +\n",
        "  my_theme +\n",
        "  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +  # Adjust x-axis limits and expansion\n",
        "  scale_x_discrete(expand = expansion(add = c(0.5, 0.5))) + # Reduce white space above and below bars\n",
        "  theme(\n",
        "          text = element_text(family = font_family),  # Set font family for the entire plot\n",
        "          plot.title = element_text(size = title_size),\n",
        "          axis.title = element_text(size = axis_title_size),\n",
        "          axis.text = element_text(size = axis_text_size),\n",
        "          axis.text.x = element_text(angle = 0, hjust=0.5),\n",
        "          axis.title.x = element_text(angle=0, hjust=0.5)\n",
        "    ) +\n",
        "  scale_fill_manual(values = color_palette) +\n",
        "  theme(legend.position = \"none\")\n",
        "\n",
        "filtered_data <- melted_data %>%\n",
        "  filter(variant == cur_ID)\n",
        "\n",
        "filtered_data <- filtered_data %>%\n",
        "  filter(cell_type %in% c(\"HEPG2\"))\n",
        "\n",
        "filtered_data$label <- ifelse(filtered_data$set == \"A\", \"Ref\", \"Alt\")\n",
        "filtered_data$label <- factor(filtered_data$label, levels = c(\"Ref\", \"Alt\"))\n",
        "\n",
        "filtered_data$cell_type_set <- interaction(filtered_data$cell_type, filtered_data$set, sep = \"_\")\n",
        "filtered_data$cell_type_set <- factor(filtered_data$cell_type_set, levels = levels(filtered_data$cell_type_set))\n",
        "\n",
        "p3<-ggplot(filtered_data, aes(x = cell_type, y = log2FC, fill = cell_type_set)) +\n",
        "  geom_bar(stat = \"identity\", position = position_dodge(width = 0.8), width = 0.7) +\n",
        "  geom_errorbar(aes(ymin = log2FC - log2FC_SE, ymax = log2FC + log2FC_SE),\n",
        "                position = position_dodge(width = 0.8), width = 0) +\n",
        "  geom_text(aes(y=-0, label = label), size=5, position = position_dodge(width = 0.8),\n",
        "            angle = 0, vjust = 0) +\n",
        "  labs(y = \"Activity - Log2(RNA/DNA)\", x = \"Cell Type\") +\n",
        "  pretty_plot() +\n",
        "  my_theme +\n",
        "  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +  # Adjust x-axis limits and expansion\n",
        "  scale_x_discrete(expand = expansion(add = c(0.5, 0.5))) + # Reduce white space above and below bars\n",
        "  theme(\n",
        "          text = element_text(family = font_family),  # Set font family for the entire plot\n",
        "          plot.title = element_text(size = title_size),\n",
        "          axis.title = element_text(size = axis_title_size),\n",
        "          axis.text = element_text(size = axis_text_size),\n",
        "          axis.text.x = element_text(angle = 0, hjust=0.5),\n",
        "          axis.title.x = element_text(angle=0, hjust=0.5)\n",
        "    ) +\n",
        "  scale_fill_manual(values = color_palette) +\n",
        "  theme(legend.position = \"none\")\n",
        "\n",
        "filtered_data <- melted_data %>%\n",
        "  filter(variant == cur_ID)\n",
        "\n",
        "filtered_data <- filtered_data %>%\n",
        "  filter(cell_type %in% c(\"K562\"))\n",
        "\n",
        "filtered_data$label <- ifelse(filtered_data$set == \"A\", \"Ref\", \"Alt\")\n",
        "filtered_data$label <- factor(filtered_data$label, levels = c(\"Ref\", \"Alt\"))\n",
        "\n",
        "filtered_data$cell_type_set <- interaction(filtered_data$cell_type, filtered_data$set, sep = \"_\")\n",
        "filtered_data$cell_type_set <- factor(filtered_data$cell_type_set, levels = levels(filtered_data$cell_type_set))\n",
        "\n",
        "p4<-ggplot(filtered_data, aes(x = cell_type, y = log2FC, fill = cell_type_set)) +\n",
        "  geom_bar(stat = \"identity\", position = position_dodge(width = 0.8), width = 0.7) +\n",
        "  geom_errorbar(aes(ymin = log2FC - log2FC_SE, ymax = log2FC + log2FC_SE),\n",
        "                position = position_dodge(width = 0.8), width = 0) +\n",
        "  geom_text(aes(y=-0, label = label), size=5, position = position_dodge(width = 0.8),\n",
        "            angle = 0, vjust = 0) +\n",
        "  labs(y = \"Activity - Log2(RNA/DNA)\", x = \"Cell Type\") +\n",
        "  pretty_plot() +\n",
        "  my_theme +\n",
        "  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +  # Adjust x-axis limits and expansion\n",
        "  scale_x_discrete(expand = expansion(add = c(0.5, 0.5))) + # Reduce white space above and below bars\n",
        "  theme(\n",
        "          text = element_text(family = font_family),  # Set font family for the entire plot\n",
        "          plot.title = element_text(size = title_size),\n",
        "          axis.title = element_text(size = axis_title_size),\n",
        "          axis.text = element_text(size = axis_text_size),\n",
        "          axis.text.x = element_text(angle = 0, hjust=0.5),\n",
        "          axis.title.x = element_text(angle=0, hjust=0.5)\n",
        "    ) +\n",
        "  scale_fill_manual(values = color_palette) +\n",
        "  theme(legend.position = \"none\")\n",
        "\n",
        "plt_combined <- p1 +  p2 + p3 / p4 + p3 + p4 + p2 + p3 + p4 + plot_annotation(title = paste0(cur_ID,\"- GTEx\")) + plot_layout(ncol=8,nrow=1, widths=c(2,1,1,1,1,2,0.5,0.5),heights=c(2,2,2,2,2,2,2,2))\n",
        "p1\n",
        "p2\n",
        "plt_combined\n",
        "print(cur_ID)\n",
        "cowplot::save_plot(\n",
        "  paste0(output_path,\"plots/\",cur_ID,\"_activity.gtex.pdf\"),\n",
        "  plt_combined,\n",
        "  base_height = 6,\n",
        "  base_width = 25,\n",
        "  device = cairo_pdf\n",
        ")\n",
        "}"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "6DSx0FWlw72U",
      "metadata": {
        "id": "6DSx0FWlw72U",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "##RUN WITH UKBB/BBJ\n",
        "\n",
        "desired_ids <- c(\"chr12:31226835:T:A\", \"chr13:50432800:T:C\", \"chr14:65175364:A:C\", \"chr15:55526886:C:T\", \"chr16:193224:G:A\", \"chr19:58662235:C:T\", \"chr19:58662254:T:C\", \"chr2:191049403:G:A\", \"chr20:45944659:G:C\", \"chr20:5485337:G:A\", \"chr20:61464184:C:T\", \"chr21:37572401:T:G\", \"chr22:19132380:A:T\", \"chr22:19132385:G:A\", \"chr22:50742721:T:A\", \"chr3:42306446:C:T\", \"chr5:95163449:G:A\", \"chr6:7143859:C:A\", \"chr7:105416560:C:A\", \"chr7:139875230:C:T\")\n",
        "\n",
        "subset_gtex <- traits_mpra_paired_final20230117 %>% filter(variant %in% desired_ids)\n",
        "\n",
        "selected_columns <- c(\"variant\",\"cell_type\",\"A_log2FC\", \"B_log2FC\", \"A_log2FC_SE\", \"B_log2FC_SE\")  # Replace with actual column names\n",
        "\n",
        "pre_filtered_data <- subset_gtex %>%\n",
        "  dplyr::select(all_of(selected_columns))\n",
        "\n",
        "#plot_data_melted <- pre_filtered_data %>%\n",
        "#  pivot_longer(\n",
        "#    cols = c(\"A_log2FC\", \"B_log2FC\"),\n",
        "#    names_to = \"log2FC_type\",\n",
        "#    values_to = \"log2FC\"\n",
        "#  )\n",
        "\n",
        "  melted_data <- pre_filtered_data %>%\n",
        "  pivot_longer(\n",
        "    cols = c(A_log2FC, A_log2FC_SE, B_log2FC, B_log2FC_SE),\n",
        "    names_to = c(\"set\", \".value\"),\n",
        "    names_pattern = \"(A|B)_(.*)\"\n",
        "  ) %>%\n",
        "  distinct()\n",
        "\n",
        "#melted_data\n",
        "\n",
        "font_family <- \"Arial\"\n",
        "title_size <- 16\n",
        "axis_title_size <- 16\n",
        "axis_text_size <- 12\n",
        "legend_title_size <- 16\n",
        "legend_text_size <- 12\n",
        "legend_key_size <- 1\n",
        "\n",
        "clrs = c(\n",
        "  pnw_palette('Bay',8,type='continuous')[5],\n",
        "  pnw_palette('Bay',8,type='continuous')[3],\n",
        "  pnw_palette('Starfish')[5],\n",
        "  pnw_palette('Bay',8,type='continuous')[1],\n",
        "  pnw_palette('Bay',8,type='continuous')[8],\n",
        "  'black'\n",
        ")\n",
        "\n",
        "names(clrs) = c(\"A549\", \"HCT116\", \"HEPG2\", \"K562\",\"SKNSH\",\"any\")\n",
        "base_colors<-clrs\n",
        "\n",
        "lighten_color <- function(color, amount = 0.3) {\n",
        "  alpha(color, amount)\n",
        "}\n",
        "\n",
        "color_palette <- c()\n",
        "for (cell_type in names(base_colors)) {\n",
        "  color_palette[paste0(cell_type, \"_A\")] <- base_colors[cell_type]\n",
        "  color_palette[paste0(cell_type, \"_B\")] <- lighten_color(base_colors[cell_type])\n",
        "}\n",
        "\n",
        "uniq_ID=unique(melted_data$variant)\n",
        "\n",
        "for (i in 1:length(uniq_ID)) {\n",
        "\n",
        "\n",
        "cur_ID=uniq_ID[i]\n",
        "print(i)\n",
        "print(cur_ID)\n",
        "\n",
        "filtered_data <- melted_data %>%\n",
        "  filter(variant == cur_ID)\n",
        "\n",
        "#print(filtered_data)\n",
        "\n",
        "filtered_data$label <- ifelse(filtered_data$set == \"A\", \"Ref\", \"Alt\")\n",
        "\n",
        "filtered_data$cell_type_set <- interaction(filtered_data$cell_type, filtered_data$set, sep = \"_\")\n",
        "filtered_data$cell_type_set <- factor(filtered_data$cell_type_set, levels = levels(filtered_data$cell_type_set))\n",
        "\n",
        "\n",
        "p1<-ggplot(filtered_data, aes(x = cell_type, y = log2FC, fill = cell_type_set)) +\n",
        "  geom_bar(stat = \"identity\", position = position_dodge(width = 0.8), width = 0.7) +\n",
        "  geom_errorbar(aes(ymin = log2FC - log2FC_SE, ymax = log2FC + log2FC_SE),\n",
        "                position = position_dodge(width = 0.8), width = 0) +\n",
        "  geom_text(aes(y=-0, label = label), size=5, position = position_dodge(width = 0.8),\n",
        "            angle = 0, vjust = 0) +\n",
        "  labs(y = \"Activity - Log2(RNA/DNA)\", x = \"Cell Type\") +\n",
        "  pretty_plot() +\n",
        "  my_theme +\n",
        "  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +  # Adjust x-axis limits and expansion\n",
        "  scale_x_discrete(expand = expansion(add = c(0.5, 0.5))) + # Reduce white space above and below bars\n",
        "  theme(\n",
        "          text = element_text(family = font_family),  # Set font family for the entire plot\n",
        "          plot.title = element_text(size = title_size),\n",
        "          axis.title = element_text(size = axis_title_size),\n",
        "          axis.text = element_text(size = axis_text_size),\n",
        "          axis.text.x = element_text(angle = 0, hjust=0.5),\n",
        "          axis.title.x = element_text(angle=0, hjust=0.5)\n",
        "    ) +\n",
        "  scale_fill_manual(values = color_palette) +\n",
        "  theme(legend.position = \"none\")\n",
        "\n",
        "filtered_data <- melted_data %>%\n",
        "  filter(variant == cur_ID)\n",
        "\n",
        "filtered_data <- filtered_data %>%\n",
        "  filter(cell_type %in% c(\"HEPG2\", \"K562\"))\n",
        "\n",
        "filtered_data$label <- ifelse(filtered_data$set == \"A\", \"Ref\", \"Alt\")\n",
        "\n",
        "filtered_data$cell_type_set <- interaction(filtered_data$cell_type, filtered_data$set, sep = \"_\")\n",
        "filtered_data$cell_type_set <- factor(filtered_data$cell_type_set, levels = levels(filtered_data$cell_type_set))\n",
        "\n",
        "p2<-ggplot(filtered_data, aes(x = cell_type, y = log2FC, fill = cell_type_set)) +\n",
        "  geom_bar(stat = \"identity\", position = position_dodge(width = 0.8), width = 0.7) +\n",
        "  geom_errorbar(aes(ymin = log2FC - log2FC_SE, ymax = log2FC + log2FC_SE),\n",
        "                position = position_dodge(width = 0.8), width = 0) +\n",
        "  geom_text(aes(y=-0, label = label), size=5, position = position_dodge(width = 0.8),\n",
        "            angle = 0, vjust = 0) +\n",
        "  labs(y = \"Activity - Log2(RNA/DNA)\", x = \"Cell Type\") +\n",
        "  pretty_plot() +\n",
        "  my_theme +\n",
        "  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +  # Adjust x-axis limits and expansion\n",
        "  scale_x_discrete(expand = expansion(add = c(0.5, 0.5))) + # Reduce white space above and below bars\n",
        "  theme(\n",
        "          text = element_text(family = font_family),  # Set font family for the entire plot\n",
        "          plot.title = element_text(size = title_size),\n",
        "          axis.title = element_text(size = axis_title_size),\n",
        "          axis.text = element_text(size = axis_text_size),\n",
        "          axis.text.x = element_text(angle = 0, hjust=0.5),\n",
        "          axis.title.x = element_text(angle=0, hjust=0.5)\n",
        "    ) +\n",
        "  scale_fill_manual(values = color_palette) +\n",
        "  theme(legend.position = \"none\")\n",
        "\n",
        "filtered_data <- melted_data %>%\n",
        "  filter(variant == cur_ID)\n",
        "\n",
        "filtered_data <- filtered_data %>%\n",
        "  filter(cell_type %in% c(\"HEPG2\"))\n",
        "\n",
        "filtered_data$label <- ifelse(filtered_data$set == \"A\", \"Ref\", \"Alt\")\n",
        "\n",
        "filtered_data$cell_type_set <- interaction(filtered_data$cell_type, filtered_data$set, sep = \"_\")\n",
        "filtered_data$cell_type_set <- factor(filtered_data$cell_type_set, levels = levels(filtered_data$cell_type_set))\n",
        "\n",
        "p3<-ggplot(filtered_data, aes(x = cell_type, y = log2FC, fill = cell_type_set)) +\n",
        "  geom_bar(stat = \"identity\", position = position_dodge(width = 0.8), width = 0.7) +\n",
        "  geom_errorbar(aes(ymin = log2FC - log2FC_SE, ymax = log2FC + log2FC_SE),\n",
        "                position = position_dodge(width = 0.8), width = 0) +\n",
        "  geom_text(aes(y=-0, label = label), size=5, position = position_dodge(width = 0.8),\n",
        "            angle = 0, vjust = 0) +\n",
        "  labs(y = \"Activity - Log2(RNA/DNA)\", x = \"Cell Type\") +\n",
        "  pretty_plot() +\n",
        "  my_theme +\n",
        "  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +  # Adjust x-axis limits and expansion\n",
        "  scale_x_discrete(expand = expansion(add = c(0.5, 0.5))) + # Reduce white space above and below bars\n",
        "  theme(\n",
        "          text = element_text(family = font_family),  # Set font family for the entire plot\n",
        "          plot.title = element_text(size = title_size),\n",
        "          axis.title = element_text(size = axis_title_size),\n",
        "          axis.text = element_text(size = axis_text_size),\n",
        "          axis.text.x = element_text(angle = 0, hjust=0.5),\n",
        "          axis.title.x = element_text(angle=0, hjust=0.5)\n",
        "    ) +\n",
        "  scale_fill_manual(values = color_palette) +\n",
        "  theme(legend.position = \"none\")\n",
        "\n",
        "filtered_data <- melted_data %>%\n",
        "  filter(variant == cur_ID)\n",
        "\n",
        "filtered_data <- filtered_data %>%\n",
        "  filter(cell_type %in% c(\"K562\"))\n",
        "\n",
        "filtered_data$label <- ifelse(filtered_data$set == \"A\", \"Ref\", \"Alt\")\n",
        "\n",
        "filtered_data$cell_type_set <- interaction(filtered_data$cell_type, filtered_data$set, sep = \"_\")\n",
        "filtered_data$cell_type_set <- factor(filtered_data$cell_type_set, levels = levels(filtered_data$cell_type_set))\n",
        "\n",
        "p4<-ggplot(filtered_data, aes(x = cell_type, y = log2FC, fill = cell_type_set)) +\n",
        "  geom_bar(stat = \"identity\", position = position_dodge(width = 0.8), width = 0.7) +\n",
        "  geom_errorbar(aes(ymin = log2FC - log2FC_SE, ymax = log2FC + log2FC_SE),\n",
        "                position = position_dodge(width = 0.8), width = 0) +\n",
        "  geom_text(aes(y=-0, label = label), size=5, position = position_dodge(width = 0.8),\n",
        "            angle = 0, vjust = 0) +\n",
        "  labs(y = \"Activity - Log2(RNA/DNA)\", x = \"Cell Type\") +\n",
        "  pretty_plot() +\n",
        "  my_theme +\n",
        "  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +  # Adjust x-axis limits and expansion\n",
        "  scale_x_discrete(expand = expansion(add = c(0.5, 0.5))) + # Reduce white space above and below bars\n",
        "  theme(\n",
        "          text = element_text(family = font_family),  # Set font family for the entire plot\n",
        "          plot.title = element_text(size = title_size),\n",
        "          axis.title = element_text(size = axis_title_size),\n",
        "          axis.text = element_text(size = axis_text_size),\n",
        "          axis.text.x = element_text(angle = 0, hjust=0.5),\n",
        "          axis.title.x = element_text(angle=0, hjust=0.5)\n",
        "    ) +\n",
        "  scale_fill_manual(values = color_palette) +\n",
        "  theme(legend.position = \"none\")\n",
        "\n",
        "plt_combined <- p1 +  p2 + p3 / p4 + p3 + p4 + p2 + p3 + p4 + plot_annotation(title = paste0(cur_ID,\"- GTEx\")) + plot_layout(ncol=8,nrow=1, widths=c(2,1,1,1,1,2,0.5,0.5),heights=c(2,2,2,2,2,2,2,2))\n",
        "p1\n",
        "p2\n",
        "plt_combined\n",
        "print(cur_ID)\n",
        "cowplot::save_plot(\n",
        "  paste0(output_path,\"plots/\",cur_ID,\"_activity.ukbb.pdf\"),\n",
        "  plt_combined,\n",
        "  base_height = 6,\n",
        "  base_width = 25,\n",
        "  device = cairo_pdf\n",
        ")\n",
        "}"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "2yQdVxtnNmz1",
      "metadata": {
        "id": "2yQdVxtnNmz1",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "head(melted_data)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "1y6TRfe5e1KO",
      "metadata": {
        "id": "1y6TRfe5e1KO",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "head(filtered_data)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "Ng3JTXGeCBb_",
      "metadata": {
        "id": "Ng3JTXGeCBb_",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "head(haplotypes_data)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "KaRO0535ixmV",
      "metadata": {
        "id": "KaRO0535ixmV",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "melted_data <- haplotypes_data\n",
        "melted_data <- melted_data %>% rename(variant = sat_ref_parent)\n",
        "melted_data <- melted_data %>% rename(log2FC = log2FC_window)\n",
        "melted_data <- melted_data %>% rename(log2FC_SE = log2FC_window_SE)\n",
        "\n",
        "#melted_data\n",
        "\n",
        "font_family <- \"Arial\"\n",
        "title_size <- 16\n",
        "axis_title_size <- 16\n",
        "axis_text_size <- 12\n",
        "legend_title_size <- 16\n",
        "legend_text_size <- 12\n",
        "legend_key_size <- 1\n",
        "\n",
        "clrs = c(\n",
        "  pnw_palette('Bay',8,type='continuous')[5],\n",
        "  pnw_palette('Bay',8,type='continuous')[3],\n",
        "  pnw_palette('Starfish')[5],\n",
        "  pnw_palette('Bay',8,type='continuous')[1],\n",
        "  pnw_palette('Bay',8,type='continuous')[8],\n",
        "  'black'\n",
        ")\n",
        "\n",
        "names(clrs) = c(\"A549\", \"HCT116\", \"HEPG2\", \"K562\",\"SKNSH\",\"any\")\n",
        "base_colors<-clrs\n",
        "\n",
        "lighten_color_AB <- function(color, amount = 0.70) {\n",
        "  alpha(color, amount)\n",
        "}\n",
        "\n",
        "lighten_color_BA <- function(color, amount = 0.40) {\n",
        "  alpha(color, amount)\n",
        "}\n",
        "\n",
        "lighten_color_BB <- function(color, amount = 0.10) {\n",
        "  alpha(color, amount)\n",
        "}\n",
        "\n",
        "color_palette <- c()\n",
        "for (cell_type in names(base_colors)) {\n",
        "  color_palette[paste0(cell_type, \"_II\")] <- base_colors[cell_type]\n",
        "  color_palette[paste0(cell_type, \"_ID\")] <- lighten_color_AB(base_colors[cell_type])\n",
        "  color_palette[paste0(cell_type, \"_DI\")] <- lighten_color_BA(base_colors[cell_type])\n",
        "  color_palette[paste0(cell_type, \"_DD\")] <- lighten_color_BB(base_colors[cell_type])\n",
        "}\n",
        "\n",
        "uniq_ID=unique(melted_data$variant)\n",
        "\n",
        "#i=1\n",
        "\n",
        "for (i in 1:length(uniq_ID)) {\n",
        "cur_ID=uniq_ID[i]\n",
        "print(i)\n",
        "print(cur_ID)\n",
        "\n",
        "filtered_data <- melted_data %>%\n",
        "  filter(variant == cur_ID)\n",
        "\n",
        "filtered_data$cell_type_set <- interaction(filtered_data$cell_type, filtered_data$allele_recode, sep = \"_\")\n",
        "filtered_data$cell_type_set <- factor(filtered_data$cell_type_set, levels = levels(filtered_data$cell_type_set))\n",
        "filtered_data$label<-filtered_data$allele\n",
        "\n",
        "p1<-ggplot(filtered_data, aes(x = cell_type, y = log2FC, fill = cell_type_set)) +\n",
        "  geom_bar(stat = \"identity\", position = position_dodge(width = 0.8), width = 0.7) +\n",
        "  geom_errorbar(aes(ymin = log2FC - log2FC_SE, ymax = log2FC + log2FC_SE),\n",
        "                position = position_dodge(width = 0.8), width = 0) +\n",
        "  geom_text(aes(y=-0, label = label), size=5, position = position_dodge(width = 0.8),\n",
        "            angle = 0, vjust = 0) +\n",
        "  labs(y = \"Activity - Log2(RNA/DNA)\", x = \"Cell Type\") +\n",
        "  pretty_plot() +\n",
        "  my_theme +\n",
        "  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +  # Adjust x-axis limits and expansion\n",
        "  scale_x_discrete(expand = expansion(add = c(0.5, 0.5))) + # Reduce white space above and below bars\n",
        "  theme(\n",
        "          text = element_text(family = font_family),  # Set font family for the entire plot\n",
        "          plot.title = element_text(size = title_size),\n",
        "          axis.title = element_text(size = axis_title_size),\n",
        "          axis.text = element_text(size = axis_text_size),\n",
        "          axis.text.x = element_text(angle = 0, hjust=0.5),\n",
        "          axis.title.x = element_text(angle=0, hjust=0.5)\n",
        "    ) +\n",
        "  scale_fill_manual(values = color_palette) +\n",
        "  theme(legend.position = \"none\")\n",
        "\n",
        "#K562 & HepG2 Only\n",
        "\n",
        "filtered_data <- melted_data %>%\n",
        "  filter(variant == cur_ID)\n",
        "\n",
        "filtered_data <- filtered_data %>%\n",
        "  filter(cell_type %in% c(\"HEPG2\", \"K562\"))\n",
        "\n",
        "filtered_data$label<-filtered_data$allele\n",
        "\n",
        "filtered_data$cell_type_set <- interaction(filtered_data$cell_type, filtered_data$allele_recode, sep = \"_\")\n",
        "filtered_data$cell_type_set <- factor(filtered_data$cell_type_set, levels = levels(filtered_data$cell_type_set))\n",
        "\n",
        "p2<-ggplot(filtered_data, aes(x = cell_type, y = log2FC, fill = cell_type_set)) +\n",
        "  geom_bar(stat = \"identity\", position = position_dodge(width = 0.8), width = 0.7) +\n",
        "  geom_errorbar(aes(ymin = log2FC - log2FC_SE, ymax = log2FC + log2FC_SE),\n",
        "                position = position_dodge(width = 0.8), width = 0) +\n",
        "  geom_text(aes(y=-0, label = label), size=5, position = position_dodge(width = 0.8),\n",
        "            angle = 0, vjust = 0) +\n",
        "  labs(y = \"Activity - Log2(RNA/DNA)\", x = \"Cell Type\") +\n",
        "  pretty_plot() +\n",
        "  my_theme +\n",
        "  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +  # Adjust x-axis limits and expansion\n",
        "  scale_x_discrete(expand = expansion(add = c(0.5, 0.5))) + # Reduce white space above and below bars\n",
        "  theme(\n",
        "          text = element_text(family = font_family),  # Set font family for the entire plot\n",
        "          plot.title = element_text(size = title_size),\n",
        "          axis.title = element_text(size = axis_title_size),\n",
        "          axis.text = element_text(size = axis_text_size),\n",
        "          axis.text.x = element_text(angle = 0, hjust=0.5),\n",
        "          axis.title.x = element_text(angle=0, hjust=0.5)\n",
        "    ) +\n",
        "  scale_fill_manual(values = color_palette) +\n",
        "  theme(legend.position = \"none\")\n",
        "\n",
        "\n",
        "### HepG2 Only\n",
        "\n",
        "filtered_data <- melted_data %>%\n",
        "  filter(variant == cur_ID)\n",
        "\n",
        "filtered_data <- filtered_data %>%\n",
        "  filter(cell_type %in% c(\"HEPG2\"))\n",
        "\n",
        "filtered_data$label<-filtered_data$allele\n",
        "\n",
        "filtered_data$cell_type_set <- interaction(filtered_data$cell_type, filtered_data$allele_recode, sep = \"_\")\n",
        "filtered_data$cell_type_set <- factor(filtered_data$cell_type_set, levels = levels(filtered_data$cell_type_set))\n",
        "\n",
        "\n",
        "p3<-ggplot(filtered_data, aes(x = cell_type, y = log2FC, fill = cell_type_set)) +\n",
        "  geom_bar(stat = \"identity\", position = position_dodge(width = 0.8), width = 0.7) +\n",
        "  geom_errorbar(aes(ymin = log2FC - log2FC_SE, ymax = log2FC + log2FC_SE),\n",
        "                position = position_dodge(width = 0.8), width = 0) +\n",
        "  geom_text(aes(y=-0, label = label), size=5, position = position_dodge(width = 0.8),\n",
        "            angle = 0, vjust = 0) +\n",
        "  labs(y = \"Activity - Log2(RNA/DNA)\", x = \"Cell Type\") +\n",
        "  pretty_plot() +\n",
        "  my_theme +\n",
        "  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +  # Adjust x-axis limits and expansion\n",
        "  scale_x_discrete(expand = expansion(add = c(0.5, 0.5))) + # Reduce white space above and below bars\n",
        "  theme(\n",
        "          text = element_text(family = font_family),  # Set font family for the entire plot\n",
        "          plot.title = element_text(size = title_size),\n",
        "          axis.title = element_text(size = axis_title_size),\n",
        "          axis.text = element_text(size = axis_text_size),\n",
        "          axis.text.x = element_text(angle = 0, hjust=0.5),\n",
        "          axis.title.x = element_text(angle=0, hjust=0.5)\n",
        "    ) +\n",
        "  scale_fill_manual(values = color_palette) +\n",
        "  theme(legend.position = \"none\")\n",
        "\n",
        "##K562 Cells Only\n",
        "\n",
        "filtered_data <- melted_data %>%\n",
        "  filter(variant == cur_ID)\n",
        "\n",
        "filtered_data <- filtered_data %>%\n",
        "  filter(cell_type %in% c(\"K562\"))\n",
        "\n",
        "filtered_data$label<-filtered_data$allele\n",
        "\n",
        "filtered_data$cell_type_set <- interaction(filtered_data$cell_type, filtered_data$allele_recode, sep = \"_\")\n",
        "filtered_data$cell_type_set <- factor(filtered_data$cell_type_set, levels = levels(filtered_data$cell_type_set))\n",
        "\n",
        "\n",
        "\n",
        "p4<-ggplot(filtered_data, aes(x = cell_type, y = log2FC, fill = cell_type_set)) +\n",
        "  geom_bar(stat = \"identity\", position = position_dodge(width = 0.8), width = 0.7) +\n",
        "  geom_errorbar(aes(ymin = log2FC - log2FC_SE, ymax = log2FC + log2FC_SE),\n",
        "                position = position_dodge(width = 0.8), width = 0) +\n",
        "  geom_text(aes(y=-0, label = label), size=5, position = position_dodge(width = 0.8),\n",
        "            angle = 0, vjust = 0) +\n",
        "  labs(y = \"Activity - Log2(RNA/DNA)\", x = \"Cell Type\") +\n",
        "  pretty_plot() +\n",
        "  my_theme +\n",
        "  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +  # Adjust x-axis limits and expansion\n",
        "  scale_x_discrete(expand = expansion(add = c(0.5, 0.5))) + # Reduce white space above and below bars\n",
        "  theme(\n",
        "          text = element_text(family = font_family),  # Set font family for the entire plot\n",
        "          plot.title = element_text(size = title_size),\n",
        "          axis.title = element_text(size = axis_title_size),\n",
        "          axis.text = element_text(size = axis_text_size),\n",
        "          axis.text.x = element_text(angle = 0, hjust=0.5),\n",
        "          axis.title.x = element_text(angle=0, hjust=0.5)\n",
        "    ) +\n",
        "  scale_fill_manual(values = color_palette) +\n",
        "  theme(legend.position = \"none\")\n",
        "\n",
        "plt_combined <- p1 +  p2 + p3 / p4 + p3 + p4 + p2 + p3 + p4 + plot_annotation(title = paste0(cur_ID,\"\")) + plot_layout(ncol=8,nrow=1, widths=c(2,1,1,1,1,2,0.5,0.5),heights=c(2,2,2,2,2,2,2,2))\n",
        "p1\n",
        "p2\n",
        "plt_combined\n",
        "print(cur_ID)\n",
        "cowplot::save_plot(\n",
        "  paste0(output_path,\"haplo_plots/\",cur_ID,\"_haplotypes_orig.pdf\"),\n",
        "  plt_combined,\n",
        "  base_height = 6,\n",
        "  base_width = 25,\n",
        "  device = cairo_pdf\n",
        ")\n",
        "}\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "XabvSTUPmuYh",
      "metadata": {
        "id": "XabvSTUPmuYh",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "head(filtered_data)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "s8cfnKdUizjK",
      "metadata": {
        "id": "s8cfnKdUizjK",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "color_palette"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "fucb6Ud-mUMl",
      "metadata": {
        "id": "fucb6Ud-mUMl",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "p4"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "6waLlcVDMDBP",
      "metadata": {
        "id": "6waLlcVDMDBP",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "melted_data <- haplotypes_data\n",
        "melted_data <- melted_data %>% rename(variant = sat_ref_parent)\n",
        "melted_data <- melted_data %>% rename(log2FC = log2FC_meta)\n",
        "melted_data <- melted_data %>% rename(log2FC_SE = log2FC_meta_SE)\n",
        "\n",
        "#melted_data\n",
        "\n",
        "font_family <- \"Arial\"\n",
        "title_size <- 16\n",
        "axis_title_size <- 16\n",
        "axis_text_size <- 12\n",
        "legend_title_size <- 16\n",
        "legend_text_size <- 12\n",
        "legend_key_size <- 1\n",
        "\n",
        "clrs = c(\n",
        "  pnw_palette('Bay',8,type='continuous')[5],\n",
        "  pnw_palette('Bay',8,type='continuous')[3],\n",
        "  pnw_palette('Starfish')[5],\n",
        "  pnw_palette('Bay',8,type='continuous')[1],\n",
        "  pnw_palette('Bay',8,type='continuous')[8],\n",
        "  'black'\n",
        ")\n",
        "\n",
        "names(clrs) = c(\"A549\", \"HCT116\", \"HEPG2\", \"K562\",\"SKNSH\",\"any\")\n",
        "base_colors<-clrs\n",
        "\n",
        "lighten_color_AB <- function(color, amount = 0.70) {\n",
        "  alpha(color, amount)\n",
        "}\n",
        "\n",
        "lighten_color_BA <- function(color, amount = 0.40) {\n",
        "  alpha(color, amount)\n",
        "}\n",
        "\n",
        "lighten_color_BB <- function(color, amount = 0.10) {\n",
        "  alpha(color, amount)\n",
        "}\n",
        "\n",
        "color_palette <- c()\n",
        "for (cell_type in names(base_colors)) {\n",
        "  color_palette[paste0(cell_type, \"_II\")] <- base_colors[cell_type]\n",
        "  color_palette[paste0(cell_type, \"_ID\")] <- lighten_color_AB(base_colors[cell_type])\n",
        "  color_palette[paste0(cell_type, \"_DI\")] <- lighten_color_BA(base_colors[cell_type])\n",
        "  color_palette[paste0(cell_type, \"_DD\")] <- lighten_color_BB(base_colors[cell_type])\n",
        "}\n",
        "\n",
        "uniq_ID=unique(melted_data$variant)\n",
        "\n",
        "#i=1\n",
        "\n",
        "for (i in 1:length(uniq_ID)) {\n",
        "cur_ID=uniq_ID[i]\n",
        "print(i)\n",
        "print(cur_ID)\n",
        "\n",
        "filtered_data <- melted_data %>%\n",
        "  filter(variant == cur_ID)\n",
        "\n",
        "filtered_data$cell_type_set <- interaction(filtered_data$cell_type, filtered_data$allele_recode, sep = \"_\")\n",
        "filtered_data$cell_type_set <- factor(filtered_data$cell_type_set, levels = levels(filtered_data$cell_type_set))\n",
        "filtered_data$label<-filtered_data$allele\n",
        "\n",
        "p1<-ggplot(filtered_data, aes(x = cell_type, y = log2FC, fill = cell_type_set)) +\n",
        "  geom_bar(stat = \"identity\", position = position_dodge(width = 0.8), width = 0.7) +\n",
        "  geom_errorbar(aes(ymin = log2FC - log2FC_SE, ymax = log2FC + log2FC_SE),\n",
        "                position = position_dodge(width = 0.8), width = 0) +\n",
        "  geom_text(aes(y=-0, label = label), size=5, position = position_dodge(width = 0.8),\n",
        "            angle = 0, vjust = 0) +\n",
        "  labs(y = \"Activity - Log2(RNA/DNA)\", x = \"Cell Type\") +\n",
        "  pretty_plot() +\n",
        "  my_theme +\n",
        "  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +  # Adjust x-axis limits and expansion\n",
        "  scale_x_discrete(expand = expansion(add = c(0.5, 0.5))) + # Reduce white space above and below bars\n",
        "  theme(\n",
        "          text = element_text(family = font_family),  # Set font family for the entire plot\n",
        "          plot.title = element_text(size = title_size),\n",
        "          axis.title = element_text(size = axis_title_size),\n",
        "          axis.text = element_text(size = axis_text_size),\n",
        "          axis.text.x = element_text(angle = 0, hjust=0.5),\n",
        "          axis.title.x = element_text(angle=0, hjust=0.5)\n",
        "    ) +\n",
        "  scale_fill_manual(values = color_palette) +\n",
        "  theme(legend.position = \"none\")\n",
        "\n",
        "#K562 & HepG2 Only\n",
        "\n",
        "filtered_data <- melted_data %>%\n",
        "  filter(variant == cur_ID)\n",
        "\n",
        "filtered_data <- filtered_data %>%\n",
        "  filter(cell_type %in% c(\"HEPG2\", \"K562\"))\n",
        "\n",
        "filtered_data$label<-filtered_data$allele\n",
        "\n",
        "filtered_data$cell_type_set <- interaction(filtered_data$cell_type, filtered_data$allele_recode, sep = \"_\")\n",
        "filtered_data$cell_type_set <- factor(filtered_data$cell_type_set, levels = levels(filtered_data$cell_type_set))\n",
        "\n",
        "p2<-ggplot(filtered_data, aes(x = cell_type, y = log2FC, fill = cell_type_set)) +\n",
        "  geom_bar(stat = \"identity\", position = position_dodge(width = 0.8), width = 0.7) +\n",
        "  geom_errorbar(aes(ymin = log2FC - log2FC_SE, ymax = log2FC + log2FC_SE),\n",
        "                position = position_dodge(width = 0.8), width = 0) +\n",
        "  geom_text(aes(y=-0, label = label), size=5, position = position_dodge(width = 0.8),\n",
        "            angle = 0, vjust = 0) +\n",
        "  labs(y = \"Activity - Log2(RNA/DNA)\", x = \"Cell Type\") +\n",
        "  pretty_plot() +\n",
        "  my_theme +\n",
        "  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +  # Adjust x-axis limits and expansion\n",
        "  scale_x_discrete(expand = expansion(add = c(0.5, 0.5))) + # Reduce white space above and below bars\n",
        "  theme(\n",
        "          text = element_text(family = font_family),  # Set font family for the entire plot\n",
        "          plot.title = element_text(size = title_size),\n",
        "          axis.title = element_text(size = axis_title_size),\n",
        "          axis.text = element_text(size = axis_text_size),\n",
        "          axis.text.x = element_text(angle = 0, hjust=0.5),\n",
        "          axis.title.x = element_text(angle=0, hjust=0.5)\n",
        "    ) +\n",
        "  scale_fill_manual(values = color_palette) +\n",
        "  theme(legend.position = \"none\")\n",
        "\n",
        "\n",
        "### HepG2 Only\n",
        "\n",
        "filtered_data <- melted_data %>%\n",
        "  filter(variant == cur_ID)\n",
        "\n",
        "filtered_data <- filtered_data %>%\n",
        "  filter(cell_type %in% c(\"HEPG2\"))\n",
        "\n",
        "filtered_data$label<-filtered_data$allele\n",
        "\n",
        "filtered_data$cell_type_set <- interaction(filtered_data$cell_type, filtered_data$allele_recode, sep = \"_\")\n",
        "filtered_data$cell_type_set <- factor(filtered_data$cell_type_set, levels = levels(filtered_data$cell_type_set))\n",
        "\n",
        "\n",
        "p3<-ggplot(filtered_data, aes(x = cell_type, y = log2FC, fill = cell_type_set)) +\n",
        "  geom_bar(stat = \"identity\", position = position_dodge(width = 0.8), width = 0.7) +\n",
        "  geom_errorbar(aes(ymin = log2FC - log2FC_SE, ymax = log2FC + log2FC_SE),\n",
        "                position = position_dodge(width = 0.8), width = 0) +\n",
        "  geom_text(aes(y=-0, label = label), size=5, position = position_dodge(width = 0.8),\n",
        "            angle = 0, vjust = 0) +\n",
        "  labs(y = \"Activity - Log2(RNA/DNA)\", x = \"Cell Type\") +\n",
        "  pretty_plot() +\n",
        "  my_theme +\n",
        "  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +  # Adjust x-axis limits and expansion\n",
        "  scale_x_discrete(expand = expansion(add = c(0.5, 0.5))) + # Reduce white space above and below bars\n",
        "  theme(\n",
        "          text = element_text(family = font_family),  # Set font family for the entire plot\n",
        "          plot.title = element_text(size = title_size),\n",
        "          axis.title = element_text(size = axis_title_size),\n",
        "          axis.text = element_text(size = axis_text_size),\n",
        "          axis.text.x = element_text(angle = 0, hjust=0.5),\n",
        "          axis.title.x = element_text(angle=0, hjust=0.5)\n",
        "    ) +\n",
        "  scale_fill_manual(values = color_palette) +\n",
        "  theme(legend.position = \"none\")\n",
        "\n",
        "##K562 Cells Only\n",
        "\n",
        "filtered_data <- melted_data %>%\n",
        "  filter(variant == cur_ID)\n",
        "\n",
        "filtered_data <- filtered_data %>%\n",
        "  filter(cell_type %in% c(\"K562\"))\n",
        "\n",
        "filtered_data$label<-filtered_data$allele\n",
        "\n",
        "filtered_data$cell_type_set <- interaction(filtered_data$cell_type, filtered_data$allele_recode, sep = \"_\")\n",
        "filtered_data$cell_type_set <- factor(filtered_data$cell_type_set, levels = levels(filtered_data$cell_type_set))\n",
        "\n",
        "\n",
        "\n",
        "p4<-ggplot(filtered_data, aes(x = cell_type, y = log2FC, fill = cell_type_set)) +\n",
        "  geom_bar(stat = \"identity\", position = position_dodge(width = 0.8), width = 0.7) +\n",
        "  geom_errorbar(aes(ymin = log2FC - log2FC_SE, ymax = log2FC + log2FC_SE),\n",
        "                position = position_dodge(width = 0.8), width = 0) +\n",
        "  geom_text(aes(y=-0, label = label), size=5, position = position_dodge(width = 0.8),\n",
        "            angle = 0, vjust = 0) +\n",
        "  labs(y = \"Activity - Log2(RNA/DNA)\", x = \"Cell Type\") +\n",
        "  pretty_plot() +\n",
        "  my_theme +\n",
        "  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +  # Adjust x-axis limits and expansion\n",
        "  scale_x_discrete(expand = expansion(add = c(0.5, 0.5))) + # Reduce white space above and below bars\n",
        "  theme(\n",
        "          text = element_text(family = font_family),  # Set font family for the entire plot\n",
        "          plot.title = element_text(size = title_size),\n",
        "          axis.title = element_text(size = axis_title_size),\n",
        "          axis.text = element_text(size = axis_text_size),\n",
        "          axis.text.x = element_text(angle = 0, hjust=0.5),\n",
        "          axis.title.x = element_text(angle=0, hjust=0.5)\n",
        "    ) +\n",
        "  scale_fill_manual(values = color_palette) +\n",
        "  theme(legend.position = \"none\")\n",
        "\n",
        "plt_combined <- p1 +  p2 + p3 / p4 + p3 + p4 + p2 + p3 + p4 + plot_annotation(title = paste0(cur_ID,\"\")) + plot_layout(ncol=8,nrow=1, widths=c(2,1,1,1,1,2,0.5,0.5),heights=c(2,2,2,2,2,2,2,2))\n",
        "p1\n",
        "p2\n",
        "plt_combined\n",
        "print(cur_ID)\n",
        "cowplot::save_plot(\n",
        "  paste0(output_path,\"haplo_plots/\",cur_ID,\"_haplotypes_meta.pdf\"),\n",
        "  plt_combined,\n",
        "  base_height = 6,\n",
        "  base_width = 25,\n",
        "  device = cairo_pdf\n",
        ")\n",
        "}"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "3V3oXLs-o8hC",
      "metadata": {
        "id": "3V3oXLs-o8hC",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "melted_data <- haplotypes_data\n",
        "melted_data <- melted_data %>% rename(variant = sat_ref_parent)\n",
        "melted_data <- melted_data %>% rename(log2FC = log2FC_satmut)\n",
        "melted_data <- melted_data %>% rename(log2FC_SE = log2FC_satmut_SE)\n",
        "\n",
        "#melted_data\n",
        "\n",
        "font_family <- \"Arial\"\n",
        "title_size <- 16\n",
        "axis_title_size <- 16\n",
        "axis_text_size <- 12\n",
        "legend_title_size <- 16\n",
        "legend_text_size <- 12\n",
        "legend_key_size <- 1\n",
        "\n",
        "clrs = c(\n",
        "  pnw_palette('Bay',8,type='continuous')[5],\n",
        "  pnw_palette('Bay',8,type='continuous')[3],\n",
        "  pnw_palette('Starfish')[5],\n",
        "  pnw_palette('Bay',8,type='continuous')[1],\n",
        "  pnw_palette('Bay',8,type='continuous')[8],\n",
        "  'black'\n",
        ")\n",
        "\n",
        "names(clrs) = c(\"A549\", \"HCT116\", \"HEPG2\", \"K562\",\"SKNSH\",\"any\")\n",
        "base_colors<-clrs\n",
        "\n",
        "lighten_color_AB <- function(color, amount = 0.70) {\n",
        "  alpha(color, amount)\n",
        "}\n",
        "\n",
        "lighten_color_BA <- function(color, amount = 0.40) {\n",
        "  alpha(color, amount)\n",
        "}\n",
        "\n",
        "lighten_color_BB <- function(color, amount = 0.10) {\n",
        "  alpha(color, amount)\n",
        "}\n",
        "\n",
        "color_palette <- c()\n",
        "for (cell_type in names(base_colors)) {\n",
        "  color_palette[paste0(cell_type, \"_II\")] <- base_colors[cell_type]\n",
        "  color_palette[paste0(cell_type, \"_ID\")] <- lighten_color_AB(base_colors[cell_type])\n",
        "  color_palette[paste0(cell_type, \"_DI\")] <- lighten_color_BA(base_colors[cell_type])\n",
        "  color_palette[paste0(cell_type, \"_DD\")] <- lighten_color_BB(base_colors[cell_type])\n",
        "}\n",
        "\n",
        "uniq_ID=unique(melted_data$variant)\n",
        "\n",
        "#i=1\n",
        "\n",
        "for (i in 1:length(uniq_ID)) {\n",
        "cur_ID=uniq_ID[i]\n",
        "print(i)\n",
        "print(cur_ID)\n",
        "\n",
        "filtered_data <- melted_data %>%\n",
        "  filter(variant == cur_ID)\n",
        "\n",
        "filtered_data$cell_type_set <- interaction(filtered_data$cell_type, filtered_data$allele_recode, sep = \"_\")\n",
        "filtered_data$cell_type_set <- factor(filtered_data$cell_type_set, levels = levels(filtered_data$cell_type_set))\n",
        "filtered_data$label<-filtered_data$allele\n",
        "\n",
        "p1<-ggplot(filtered_data, aes(x = cell_type, y = log2FC, fill = cell_type_set)) +\n",
        "  geom_bar(stat = \"identity\", position = position_dodge(width = 0.8), width = 0.7) +\n",
        "  geom_errorbar(aes(ymin = log2FC - log2FC_SE, ymax = log2FC + log2FC_SE),\n",
        "                position = position_dodge(width = 0.8), width = 0) +\n",
        "  geom_text(aes(y=-0, label = label), size=5, position = position_dodge(width = 0.8),\n",
        "            angle = 0, vjust = 0) +\n",
        "  labs(y = \"Activity - Log2(RNA/DNA)\", x = \"Cell Type\") +\n",
        "  pretty_plot() +\n",
        "  my_theme +\n",
        "  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +  # Adjust x-axis limits and expansion\n",
        "  scale_x_discrete(expand = expansion(add = c(0.5, 0.5))) + # Reduce white space above and below bars\n",
        "  theme(\n",
        "          text = element_text(family = font_family),  # Set font family for the entire plot\n",
        "          plot.title = element_text(size = title_size),\n",
        "          axis.title = element_text(size = axis_title_size),\n",
        "          axis.text = element_text(size = axis_text_size),\n",
        "          axis.text.x = element_text(angle = 0, hjust=0.5),\n",
        "          axis.title.x = element_text(angle=0, hjust=0.5)\n",
        "    ) +\n",
        "  scale_fill_manual(values = color_palette) +\n",
        "  theme(legend.position = \"none\")\n",
        "\n",
        "#K562 & HepG2 Only\n",
        "\n",
        "filtered_data <- melted_data %>%\n",
        "  filter(variant == cur_ID)\n",
        "\n",
        "filtered_data <- filtered_data %>%\n",
        "  filter(cell_type %in% c(\"HEPG2\", \"K562\"))\n",
        "\n",
        "filtered_data$label<-filtered_data$allele\n",
        "\n",
        "filtered_data$cell_type_set <- interaction(filtered_data$cell_type, filtered_data$allele_recode, sep = \"_\")\n",
        "filtered_data$cell_type_set <- factor(filtered_data$cell_type_set, levels = levels(filtered_data$cell_type_set))\n",
        "\n",
        "p2<-ggplot(filtered_data, aes(x = cell_type, y = log2FC, fill = cell_type_set)) +\n",
        "  geom_bar(stat = \"identity\", position = position_dodge(width = 0.8), width = 0.7) +\n",
        "  geom_errorbar(aes(ymin = log2FC - log2FC_SE, ymax = log2FC + log2FC_SE),\n",
        "                position = position_dodge(width = 0.8), width = 0) +\n",
        "  geom_text(aes(y=-0, label = label), size=5, position = position_dodge(width = 0.8),\n",
        "            angle = 0, vjust = 0) +\n",
        "  labs(y = \"Activity - Log2(RNA/DNA)\", x = \"Cell Type\") +\n",
        "  pretty_plot() +\n",
        "  my_theme +\n",
        "  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +  # Adjust x-axis limits and expansion\n",
        "  scale_x_discrete(expand = expansion(add = c(0.5, 0.5))) + # Reduce white space above and below bars\n",
        "  theme(\n",
        "          text = element_text(family = font_family),  # Set font family for the entire plot\n",
        "          plot.title = element_text(size = title_size),\n",
        "          axis.title = element_text(size = axis_title_size),\n",
        "          axis.text = element_text(size = axis_text_size),\n",
        "          axis.text.x = element_text(angle = 0, hjust=0.5),\n",
        "          axis.title.x = element_text(angle=0, hjust=0.5)\n",
        "    ) +\n",
        "  scale_fill_manual(values = color_palette) +\n",
        "  theme(legend.position = \"none\")\n",
        "\n",
        "\n",
        "### HepG2 Only\n",
        "\n",
        "filtered_data <- melted_data %>%\n",
        "  filter(variant == cur_ID)\n",
        "\n",
        "filtered_data <- filtered_data %>%\n",
        "  filter(cell_type %in% c(\"HEPG2\"))\n",
        "\n",
        "filtered_data$label<-filtered_data$allele\n",
        "\n",
        "filtered_data$cell_type_set <- interaction(filtered_data$cell_type, filtered_data$allele_recode, sep = \"_\")\n",
        "filtered_data$cell_type_set <- factor(filtered_data$cell_type_set, levels = levels(filtered_data$cell_type_set))\n",
        "\n",
        "\n",
        "p3<-ggplot(filtered_data, aes(x = cell_type, y = log2FC, fill = cell_type_set)) +\n",
        "  geom_bar(stat = \"identity\", position = position_dodge(width = 0.8), width = 0.7) +\n",
        "  geom_errorbar(aes(ymin = log2FC - log2FC_SE, ymax = log2FC + log2FC_SE),\n",
        "                position = position_dodge(width = 0.8), width = 0) +\n",
        "  geom_text(aes(y=-0, label = label), size=5, position = position_dodge(width = 0.8),\n",
        "            angle = 0, vjust = 0) +\n",
        "  labs(y = \"Activity - Log2(RNA/DNA)\", x = \"Cell Type\") +\n",
        "  pretty_plot() +\n",
        "  my_theme +\n",
        "  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +  # Adjust x-axis limits and expansion\n",
        "  scale_x_discrete(expand = expansion(add = c(0.5, 0.5))) + # Reduce white space above and below bars\n",
        "  theme(\n",
        "          text = element_text(family = font_family),  # Set font family for the entire plot\n",
        "          plot.title = element_text(size = title_size),\n",
        "          axis.title = element_text(size = axis_title_size),\n",
        "          axis.text = element_text(size = axis_text_size),\n",
        "          axis.text.x = element_text(angle = 0, hjust=0.5),\n",
        "          axis.title.x = element_text(angle=0, hjust=0.5)\n",
        "    ) +\n",
        "  scale_fill_manual(values = color_palette) +\n",
        "  theme(legend.position = \"none\")\n",
        "\n",
        "##K562 Cells Only\n",
        "\n",
        "filtered_data <- melted_data %>%\n",
        "  filter(variant == cur_ID)\n",
        "\n",
        "filtered_data <- filtered_data %>%\n",
        "  filter(cell_type %in% c(\"K562\"))\n",
        "\n",
        "filtered_data$label<-filtered_data$allele\n",
        "\n",
        "filtered_data$cell_type_set <- interaction(filtered_data$cell_type, filtered_data$allele_recode, sep = \"_\")\n",
        "filtered_data$cell_type_set <- factor(filtered_data$cell_type_set, levels = levels(filtered_data$cell_type_set))\n",
        "\n",
        "\n",
        "\n",
        "p4<-ggplot(filtered_data, aes(x = cell_type, y = log2FC, fill = cell_type_set)) +\n",
        "  geom_bar(stat = \"identity\", position = position_dodge(width = 0.8), width = 0.7) +\n",
        "  geom_errorbar(aes(ymin = log2FC - log2FC_SE, ymax = log2FC + log2FC_SE),\n",
        "                position = position_dodge(width = 0.8), width = 0) +\n",
        "  geom_text(aes(y=-0, label = label), size=5, position = position_dodge(width = 0.8),\n",
        "            angle = 0, vjust = 0) +\n",
        "  labs(y = \"Activity - Log2(RNA/DNA)\", x = \"Cell Type\") +\n",
        "  pretty_plot() +\n",
        "  my_theme +\n",
        "  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +  # Adjust x-axis limits and expansion\n",
        "  scale_x_discrete(expand = expansion(add = c(0.5, 0.5))) + # Reduce white space above and below bars\n",
        "  theme(\n",
        "          text = element_text(family = font_family),  # Set font family for the entire plot\n",
        "          plot.title = element_text(size = title_size),\n",
        "          axis.title = element_text(size = axis_title_size),\n",
        "          axis.text = element_text(size = axis_text_size),\n",
        "          axis.text.x = element_text(angle = 0, hjust=0.5),\n",
        "          axis.title.x = element_text(angle=0, hjust=0.5)\n",
        "    ) +\n",
        "  scale_fill_manual(values = color_palette) +\n",
        "  theme(legend.position = \"none\")\n",
        "\n",
        "plt_combined <- p1 +  p2 + p3 / p4 + p3 + p4 + p2 + p3 + p4 + plot_annotation(title = paste0(cur_ID,\"\")) + plot_layout(ncol=8,nrow=1, widths=c(2,1,1,1,1,2,0.5,0.5),heights=c(2,2,2,2,2,2,2,2))\n",
        "p1\n",
        "p2\n",
        "plt_combined\n",
        "print(cur_ID)\n",
        "cowplot::save_plot(\n",
        "  paste0(output_path,\"haplo_plots/\",cur_ID,\"_haplotypes_SatMut.pdf\"),\n",
        "  plt_combined,\n",
        "  base_height = 6,\n",
        "  base_width = 25,\n",
        "  device = cairo_pdf\n",
        ")\n",
        "}"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "private_outputs": true,
      "provenance": []
    },
    "kernelspec": {
      "display_name": "R",
      "name": "ir"
    },
    "language_info": {
      "name": "R"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}
